### DISCUSSION TITLE
MDLX major refactor

### DISCUSSION THREAD

**Discussion by Xeeynamo** [OpenKH/OpenKh#622] [Ideas]
# MDLX major refactor

## Why

The current `Mdlx` implementation is fundamentally badly designed.

A MDLX is a file extension that is internally stored as a BAR file. yet the `Kh2.Mdlx` parser does operate on a file within a MDLX file. It does not make much sense. The `Kh2.Mdlx` really is a model parser.

Kingdom Hearts II has different model types:

* Legacy: Those are Kingdom Hearts 1 models. KH2 lacks of logic to read them but the identifier is there.
* Background: models that do not represent an entity, like maps
* Skeleton: models that can be animated
* Shadow: models that are flattered on the ground and rendered in their own special way
* Gumi: gummiship models
* Multi: A model that contains more models within it. At the moment I am not aware of any model that falls under this category

Each of them contains their own fields, flags and properties and they share very little in common. Yet our current `Kh2.Mdlx` tries to do everything. This lead the introduction of `Kh2.Mdlx.IsEntity` and `Kh2.Mdlx.IsMap` to circumnavigate the bad design of the parser.

## How

Separate each model type into their own class (eg. `ModelLegacy`, `ModelBackground`) and let them implement the abstracted class `Model`.

```csharp
public abstract class Model
{
    public static Model Read(Stream);
    public void Write(Stream);
    public int GetDrawPolygonCount(IList<byte> displayFlags);
    public IEnumerable<DrawChain> GetDrawChain(IList<byte> displayFlags);
}
```

`Model.Read` and `Model.Write` will parse and write a model file as usual, but internally they will use the appropriate implementation based on the model type.

 `Model` needs to have virtual methods such as `GetDrawPolygonCount` and `GetDrawChain`. The method `GetDrawChain` will be extremely important as it will completely abstract from the draw engine how models are supposed to be rendered. The draw engine will just get a draw chain from the model and render it, unaware of the underlying implementation.

My suggestion is to slowly transition to the new format. Create `Model` and the underlying model types and wrap them into the existing `Mdlx` parser. Once `Mdlx` becomes no more than an empty shell, delete it and expose `Model` directly.

**Comment by osdanova** [comment]
Are there examples of said types? There should only be "Skeleton", "Background" and "Multi" afaik.
I agree with the rest.

**Reply by Xeeynamo** [reply]
Shadow is usually found inside the MDLX in the OBJ folder. I think it's right after the ModelSkeleton?. Gumi seems to be a subset of KH2 logic by its own and apparently the model format is slightly different. Legacy is basically KH1, which shares a lot with KH2.

**Reply by kenjiuno** [reply]
This is a little nosy though, it shows relation with Skeleton and Shadow. Model is chained.

![2022-08-15_19h45_40](https://user-images.githubusercontent.com/5955540/184622236-9ad08f6c-1860-4be0-b2c9-4a3e7098a01c.png)

If you are trying to look up this,  visit `Kaitai Web IDE`, and import the following ksy file as a plain text, and drop a mdlx/map file for parsing.

- https://ide.kaitai.io/
- https://gitlab.com/kenjiuno/khkh_xldM/-/blob/master/Parser/mdlx/bar-model-v2.ksy

**Reply by kenjiuno** [reply]
entity model p_ex100.mdlx

![image](https://user-images.githubusercontent.com/5955540/184791495-6e4df162-e959-4d44-ae4f-c5c87ffc307d.png)

map model tt03.map

![image](https://user-images.githubusercontent.com/5955540/184792135-f661b6aa-0237-494f-bbad-476a5d995ecd.png)

**Comment by kenjiuno** [comment]
I wrote down public modifier parts of current structure. It is about mdlx parsers.
It is clear that MdlxParser will be better to have `Model` instead of `Mdlx`.

![MDLX parser](https://user-images.githubusercontent.com/5955540/184767093-27bd8207-06d8-4e8e-96db-d8c2a10d0188.png)
